tokio/src/task/core.rs

```diff
@@ -55,102 +54,213 @@ pub(crate) struct Header {
+                    _ => unreachable!("unexpected stage"),
+                };
 
-            // The future is pinned within the task. The above state transition
-            // has ensured the safety of this action.
-            let future = unsafe { Pin::new_unchecked(future) };
+                // Safety: The caller ensures the future is pinned.
+                let future = unsafe { Pin::new_unchecked(future) };
 
-            // The waker passed into the `poll` function does not require a ref
-            // count increment.
-            let waker_ref = waker_ref::<T, S>(header);
-            let mut cx = Context::from_waker(&*waker_ref);
+                // The waker passed into the `poll` function does not require a ref
+                // count increment.
+                let waker_ref = waker_ref::<T, S>(header);
+                let mut cx = Context::from_waker(&*waker_ref);
 
-            future.poll(&mut cx)
+                future.poll(&mut cx)
+            })
         };
 
         if res.is_ready() {
-            self.stage = Stage::Consumed;
+            self.drop_future_or_output();
         }
 
         res
     }
 
-    pub(super) fn store_output(&mut self, output: super::Result<T::Output>) {
-        self.stage = Stage::Finished(Track::new(output));
+    /// Drop the future
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure it is safe to mutate the `stage` field.
+    pub(super) fn drop_future_or_output(&self) {
+        self.stage.with_mut(|ptr| {
+            // Safety: The caller ensures mutal exclusion to the field.
+            unsafe { *ptr = Stage::Consumed };
+        });
+    }
+
+    /// Store the task output
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure it is safe to mutate the `stage` field.
+    pub(super) fn store_output(&self, output: super::Result<T::Output>) {
+        self.stage.with_mut(|ptr| {
+            // Safety: the caller ensures mutal exclusion to the field.
+            unsafe { *ptr = Stage::Finished(output) };
+        });
     }
 
-    pub(super) unsafe fn read_output(&mut self, dst: *mut Track<super::Result<T::Output>>) {
+    /// Take the task output
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure it is safe to mutate the `stage` field.
+    pub(super) fn take_output(&self) -> super::Result<T::Output> {
         use std::mem;
 
-        dst.write(match mem::replace(&mut self.stage, Stage::Consumed) {
-            Stage::Finished(output) => output,
-            _ => unreachable!("unexpected state"),
+        self.stage.with_mut(|ptr| {
+            // Safety:: the caller ensures mutal exclusion to the field.
+            match mem::replace(unsafe { &mut *ptr }, Stage::Consumed) {
+                Stage::Finished(output) => output,
+                _ => panic!("unexpected task state"),
+            }
+        })
+    }
+
+    /// Schedule the future for execution
+    pub(super) fn schedule(&self, task: Notified<S>) {
```

Comment:
Since this is `pub(super)`, I think it should still be `unsafe fn`.

Reply:
The safety boundary is `task`. I agree that it is somewhat confusing because `LocalSet` is located there as well. After this PR, the task "'internals" will be moved to `runtime::task` and that will be the safety boundary.